### PR TITLE
CORE-2033 Add RegistrationV4 contract

### PR DESCRIPTION
**Pull request by kaihirota**
# Summary

# Why the changes

# Things worth calling out

# Before merging
- [ ] ***For Immutable developers:*** Do the changes in this PR require documentation updates? Please refer to this [SDK documentation guide](https://immutable.atlassian.net/wiki/spaces/PPS/pages/1916994017/SDK+documentation+guide) and list links to documentation update PRs (they don't have to be merged) below (or write `N/A`):
    - [ ] *Add documentation update 1*
    - [ ] *Add documentation update 2*

### CODE REVIEW

File: src/contracts/contracts/RegistrationV2.ts [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+import type { FunctionFragment, Result } from "@ethersproject/abi";
+import type { Listener, Provider } from "@ethersproject/providers";
+import type {
+  TypedEventFilter,
+  TypedEvent,
+  TypedListener,
+  OnEvent,
+} from "../common";
+
+export interface RegistrationV2Interface extends utils.Interface {
```
Comment (brayansdt) [inline src/contracts/contracts/RegistrationV2.ts:25]: Can we please rename it to RegistrationV4 as the contract name deployed on-chain is now called RegistrationV4?

**Review by carleeto — APPROVED** [review]

File: contracts/v4/Core.sol [hunk 2]
```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
+
+interface Core {
```
Comment (brayansdt) [inline contracts/v4/Core.sol:4]: this is outdated, can you please have a look at https://github.com/immutable/contracts/tree/main/contracts/bridge/x/v4

File: contracts/v4/Registration.sol [hunk 3]
```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+import {Core} from "./Core.sol";
+
+contract Registration {
```
Comment (brayansdt) [inline contracts/v4/Registration.sol:6]: this is outdated, can you please have a look at https://github.com/immutable/contracts/tree/main/contracts/bridge/x/v4

File: contracts/v4/CoreV4.sol [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    function lockVault(uint256 assetId, uint256 vaultId, uint256 lockTime) external;
+
+    function setDefaultVaultWithdrawalLock(uint256 newDefaultTime) external;
+
+    function withdrawFromVault(uint256 assetId, uint256 vaultId, uint256 quantizedAmount) external;
+
+    event ImplementationActivationRescheduled(address indexed implementation, uint256 updatedActivationTime);
+
+    function updateImplementationActivationTime(address implementation, bytes memory data, bool finalize) external;
+}
```
Comment (brayansdt) [inline contracts/v4/CoreV4.sol:277]:
```suggestion
}

```

File: contracts/v4/RegistrationV4.sol [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    function getVersion() external view returns (string memory) {
+        return imx.VERSION();
+    }
+
+    function isRegistered(uint256 starkKey) public view returns (bool) {
+        return imx.getEthKey(starkKey) != address(0);
+    }
+}
```
Comment (brayansdt) [inline contracts/v4/RegistrationV4.sol:81]:
```suggestion
}

```

File: src/contracts/contracts/RegistrationV4.ts [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    ): Promise<PopulatedTransaction>;
+
+    withdrawAll(
+      ethKey: BigNumberish,
+      starkKey: BigNumberish,
+      assetType: BigNumberish,
+      overrides?: Overrides & { from?: string }
+    ): Promise<PopulatedTransaction>;
+  };
+}
```
Comment (brayansdt) [inline src/contracts/contracts/RegistrationV4.ts:344]:
```suggestion
}

```

File: src/contracts/factories/contracts/RegistrationV4__factory.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+  static createInterface(): RegistrationV4Interface {
+    return new utils.Interface(_abi) as RegistrationV4Interface;
+  }
+  static connect(
+    address: string,
+    signerOrProvider: Signer | Provider
+  ): RegistrationV4 {
+    return new Contract(address, _abi, signerOrProvider) as RegistrationV4;
+  }
+}
```
Comment (brayansdt) [inline src/contracts/factories/contracts/RegistrationV4__factory.ts]:
```suggestion
}

```

**Review by brayansdt — APPROVED** [review]

**Pull request merged**